Add Polarity Invert option to Power sensors in Energy dashboard

TCWORLD opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

In #27373 the option to track power as well as energy was added. In the case of the Grid and Battery power sources, these are added as a single bi-directional sensor which requires a specific polarity to be meaningful. However not all integrations or hardware use the expected polarity meaning additional template sensors would need to be created to negate these values before they could be used with the energy dashboard.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->

In this PR, and the associated core PR, we add the option to invert the polarity of a power sensor used for grid/battery/solar power data. This is saved in the preferences as stat_rate_negate alongside with the stat_rate entity id.

In the dialog entries, a check-box is added to select whether to invert the polarity.

<img width="300" alt="Battery power with checked invert option" src="https://github.com/user-attachments/assets/589d53e9-0177-4d23-947d-ccb1fbb2e390" />
<img width="300" alt="Solar power with unchecked invert option" src="https://github.com/user-attachments/assets/47a6f185-5d99-451e-89a7-01469cd30b69" />
<img width="300" alt="Grid power with unchecked invert option" src="https://github.com/user-attachments/assets/e9140c13-544c-4b5a-8cdc-41c7cd890f12" />

The scope was limited to these two sources as they are likely to be bidirectional. Solar and individual loads are unlikely to need inverting, though that could be added at a later date.

## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Core pull request: https://github.com/home-assistant/core/pull/158142
- This PR fixes or closes issue: 
  #28409 
  https://github.com/wills106/homeassistant-solax-modbus/issues/1751
  https://github.com/WillCodeForCats/solaredge-modbus-multi/issues/927
- This PR is related to issue or discussion: https://github.com/orgs/home-assistant/discussions/1915
- Link to documentation pull request:

There was a similar proposal to add a direction toggle in the comments of the original PR: https://github.com/home-assistant/core/pull/153809#issuecomment-3401538017

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

TCWORLD committed:
Add stat_negate to Grid/Battery Power Preferences
This is a boolean flag which will be used to indicate whether a bi-directional power sensor should be negated if a negative value indicates the wrong direction to expected.

TCWORLD committed:
Add data handling for stat_negate option
When there is a negate preference present and it is set to true, the value for power retrieved from the entity will be negated (value = 0 - value).

TCWORLD committed:
Add Polarity Invert checkbox to the Battery and Grid power dialogs.

The new form field configures whether a power meter has inverted polarity.

TCWORLD committed:
Merge branch 'dev' into power-invert-opt

TCWORLD committed:
Merge branch 'dev' into power-invert-opt

karwosts commented:
> Can you also do this for Solar? My kWh meter also gives negative values

Suggest to wait for approval of the current before adding anything more. Can always go back and add more when/if the concept is generally approved.

TheFes commented:
Maybe add the checkbox for both battery power sensors. There are integrations which reverse the logic HA uses, so they are negative when HA expects positive, and the other way around. 

Anyway, very nice that you're adding this. Thanks!

Edit: based on the comment above, maybe wait until this one is merged and add that as a follow up :)

TCWORLD commented:
@pergolafabio This PR is only dealing with the power sensors which use a specific function to access the data. Energy sensors would be possibly more complex, and so ought to be a different FR.

@TheFes There's only one power sensor for the battery as far as I can tell. The other two are energy sensors.

pergolafabio commented:
> @pergolafabio This PR is only dealing with the power sensors which use a specific function to access the data. Energy sensors would be possibly more complex, and so ought to be a different FR.
> 
> @TheFes There's only one power sensor for the battery as far as I can tell. The other two are energy sensors.

No. Im talking about power sensors, I use a KWh meter for my solar, the kwh meter reports also negative... I had to create a template entity to add solar as power source to make it positive

TheFes commented:
@pergolafabio kWh is energy, power sensors will be using (k)W.
As to keep referencing kWh, you are causing confusion :)

pergolafabio commented:
I use this integration, that power meter reports negative   values...
<img width="1080" height="2400" alt="Screenshot_20251207-204430" src="https://github.com/user-attachments/assets/aaa2203f-8dd0-4286-80f0-b31bbfc5e9c0" />

pergolafabio commented:
> @pergolafabio Ah ok, I was confused as kWh is energy unit.
> 
> It should be easy enough to add the checkbox for the solar power sensor as well.

Would be great, thnx

TCWORLD committed:
Rename config stat_negate to stat_rate_negate
To make it clearer what entity it is related to.

TCWORLD committed:
Improve power invert label text
Sensor doesn't have to be bi-directional to invert it.

TCWORLD committed:
Add Missing Invert of Battery Power
The negate preference had not being used when getting the battery data.

TCWORLD committed:
Add Power Negate option to Solar Power Sensor
Some exotic setups may have a negative power value for solar. Given grid and battery have negate options it doesn't seem out of place for solar to have that also.

WillCodeForCats commented:
Thank you for this. SolarEdge inverters with an Export+Import meter provide a negative power value when importing from grid and a positive power value when exporting to grid (positive means excess solar generation).

I'd already created a new inverted meter power sensor entity in [solaredge-modbus-multi](https://github.com/WillCodeForCats/solaredge-modbus-multi) to provide the polarity HA wanted from the integration so that users aren't forced to create template sensors, but I'll hold off on that for a bit.

MindFreeze reviewed (commented):
I know this would be very useful but I don't think this is the way to do it. It adds a lot of complexity and fragility to the data handling. I already see that you don't properly handle the polarity of statistics data in the power sources graph, just the latest value, and these kinds of bugs will just keep multiplying every time we add something new about power data.
A more robust approach would be to create new sensors in the energy integration in core. That basically act as the template sensor you would normally create yourself but handled internally.

TCWORLD committed:
Fix statistics handling for negating power sensors

NoRi2909 commented on src/translations/en.json:
```diff
@@ -3112,7 +3112,8 @@
               "header": "Configure grid power",
               "power_stat": "Power sensor",
-              "power_helper": "Pick a sensor which measures grid power in either of {unit}. Positive values indicate importing electricity from the grid, negative values indicate exporting electricity to the grid."
+              "power_negate": "Invert power sensor polarity",
```
"negate" as the key name is misleading, should be "invert" instead.

Same for the other two below which instead could be replaced with a reference, too.

TCWORLD committed:
Invert power direction in getPowerFromState()

For consistent access to the "now" value, put the polarity invert back into getPowerFromState().

WillCodeForCats commented:
Existing sensors should be used because they have a lifetime of long term stats behind them.

MindFreeze commented:
And as the self appointed Principal Chart Developer of OHF (mostly a joke), I can tell you right now that if I have to account for every possible way to log power data every time I work on a chart, I just won't work on it at all. It's complicated enough as it is.

TCWORLD commented:
I can move the above post to https://github.com/orgs/home-assistant/discussions/1915 if it is easier to continue this discussion there.
